modelmapper/fetcher.py

```diff
@@ -0,0 +1,262 @@
+try:
+    import boto3
+except ImportError:
+    class boto3:
+        def client(self, *args, **kwargs):
+            raise ImportError('Please install Boto3')
+
+try:
+    import mmh3
+except ImportError:
+    class mmh3:
+        def hash(self, *args, **kwargs):
+            raise ImportError('Please install mmh3')
+
+import gzip
+import datetime
+import logging
+
+from collections import Mapping
+from botocore.client import Config
+
+from modelmapper.cleaner import Cleaner
+from modelmapper.slack import slack
+from modelmapper.misc import generator_chunker, generator_updater
+
+try:
+    from sqlalchemy.dialects.postgresql import insert
+except ImportError:
+    def insert(table):
+        raise ImportError('Please install SQLAlchemy')
+
+
+class Fetcher(Cleaner):
```

Comment:
If I recall correctly, cleaner extends mapper, right?  I would try to design this so Fetcher is an argument to the Cleaner/Mapper constructor, and not a class extension of it.

I'd also call this S3Fetcher since it interfaces directly to S3.  I'd create another one for reading from the file system, and another for postgres, etc.

Reply:
Or, I'd make both the mapper and the fetcher take in a cleaner if I'm not quite understanding how Mapper, Fetcher and Cleaner are releated

Reply:
@JackWink Mapper includes the settings loading, csv loading and model training functionality. Cleaner subclasses Mapper and adds data cleaning functionality to it. Fetcher subclasses cleaner and adds data insertion to the database functionality and it adds the functionality to be able to use S3 buckets. You don't have to use the s3 functionality there. It is just there for you to use it if you want it. I will put it in a mixin though to make that more clear. As far as Fetcher's point of view, `get_client_data` which will be defined by the user of this class, will provide the data that then Fetcher will use to import. The only assumption regarding S3 that is makes is that you want to put a backup of the original raw data that `get_client_data` gives you, on S3 bucket. But I will remove that assumption and leave it up to the user so it is decoupled.

Reply:
Right, but I'd still suggest making `Cleaner` an argument you can pass into the `Mapper`, and use the `Cleaner` from within `Mapper`, rather than having `Cleaner` extend `Mapper`. Similarly with the `Fetcher`. It lets you test the behaviors of each in isolation better, and makes it easier to continue to extend the functionality of the `Mapper`.